src/Microsoft.AspNetCore.SignalR.Client/HubConnection.cs

```diff
@@ -219,19 +219,24 @@ private void Shutdown(Exception ex = null)
 
             _connectionActive.Cancel();
 
-            lock (_pendingCallsLock)
+            // Keys is a snapshot, so we'll need to run this until other threads stop adding (which they will because of the cancellation token above)
```

Comment:
Isn't it racy though? If you expect/allow other threads to add while you are in this loop the threads can be scheduled in a way that _pendingCalls.Count is 0 so you will exit the loop after which a new item will be added.